Rewrote the config module to be more flexible and isolated the clap dependency into the common module

soenkeliebau opened the pull request:
…

soenkeliebau committed:
Rewrote the config module to be more flexible and isolated the clap dependency into the common module.

soenkeliebau committed:
Fixed RustFMT complaints.

soenkeliebau committed:
Some additional comments.

lfrancke commented on config/src/lib.rs:
```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
+        // Parse command line
+        let test = commandline.clone().expect("ie");
+        println!("Commandline: {:?}", test);
+        let matcher = matcher.get_matches_from(commandline.expect("trnsi"));
```
...test leftover stuff

soenkeliebau replied:
Removed

lfrancke commented on config/src/lib.rs:
```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
+struct ConfigBuilder {}
+
+impl ConfigBuilder {
+    /// The main function in
```
Missing stuff

soenkeliebau replied:
Fixed

lfrancke commented on config/src/lib.rs:
```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
+        // -> ParsedValues
+        let description = T::get_config_description();
+
+        //
```
Empty comment

soenkeliebau replied:
Fixed

fwaibel commented on config/src/lib.rs:
```diff
@@ -180,49 +299,112 @@ mod tests {
+
+        // Helper function to check whether the argument was provided on the command line
+        pub fn argument_was_provided(&self, key: &ConfigOption) -> bool {
+            let test = self.values.get(key).unwrap();
```
This line seems to be debug statement

soenkeliebau replied:
Correct, removed it.

fwaibel commented on config/src/lib.rs:
```diff
@@ -180,49 +299,112 @@ mod tests {
+                .expect("Fatal error: key not present in HashMap, but should have been!")
+            {
+                None => false,
+                Some(v) => true,
```
Could we replace the Some(v) with the wildcard "_"?
`_ => true,`

soenkeliebau replied:
Indeed we could. I've rewritten the entire statement a little as an "if let" block instead of match, which seems to be the recommended way - even though I'm not sure I agree that it is more readable :)

fwaibel commented on config/src/lib.rs:
```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
+            .about(config.about);
+
+        for option in config.options.iter() {
+            let mut new_arg = Arg::with_name(option.name)
```
The compiler issues a warning for this line. I wonder why we need a `mut`able variable here.

soenkeliebau replied:
Mental error on my part, I added the _mut_, when I refactored the code a few lines further down to conditionally allow multiple arguments or not. 
I originally had the "if option.list {" part as mutating the new_arg variable and at that time needed the mut.

Removed it.

fwaibel reviewed (commented):
Just my 2 ¢...

lfrancke reviewed (changes requested):
Sorry, not finished the review yet, submitting anyway. Will continue later.

lfrancke commented on config/src/lib.rs:
```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
+        // Convert results from command line parsing into a HashMap<ConfigOption, Vec<String>>
+        // this is then passed to the actual implementation of the configuration for processing
+        let mut result: HashMap<ConfigOption, Option<Vec<String>>> = HashMap::new();
+        for config_option in description.options.clone() {
```
This clone() is not needed when the `create_matcher` method gets proper lifetime annotations.
I'm not suggesting we necessarily do this.

https://users.rust-lang.org/t/borrow-later-used-here-i-dont-understand/49886

soenkeliebau replied:
I think we can leave the clone in this case, shouldn't be too expensive and in case it creates issues down the line we can revisit this.

lfrancke commented on config/src/lib.rs:
```diff
@@ -56,106 +80,199 @@ pub struct ConfigOption {
+        // this is then passed to the actual implementation of the configuration for processing
+        let mut result: HashMap<ConfigOption, Option<Vec<String>>> = HashMap::new();
+        for config_option in description.options.clone() {
+            println!(
```
Is this also debug stuff?

soenkeliebau replied:
Indeed

soenkeliebau committed:
Addressed comments from Lars & Florian

lfrancke commented on config/src/lib.rs:
```diff
@@ -56,106 +80,201 @@ pub struct ConfigOption {
+            if matcher.occurrences_of(config_option.name) != 0 {
+                let found_values = matcher
+                    .values_of(config_option.name)
+                    .expect("error")
```
More descriptive error message?
I know it shouldn't happen because you already checked if it's available but maybe change to at least "Should never happen" or something like that.

soenkeliebau replied:
Added proper error message.

lfrancke commented on config/src/lib.rs:
```diff
@@ -180,49 +301,111 @@ mod tests {
-            TestConfig::TEST_PARAM2.default
-        )
+            config.argument_was_provided(&TestConfig::TEST_PARAM2),
+            false
```
You removed the check whether the default value is used. I think that'd be a useful test to have.

soenkeliebau replied:
Added those checks back in.

soenkeliebau committed:
Addressed comments from Lars & Florian

Added handling for default values so that these are parsed and populated by clap.

Added a test case to check that default values are set correctly.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/stackabletech/common?pullRequest=6) <br/>All committers have signed the CLA.
